userspace/libscap/scap_savefile.c

```diff
@@ -259,6 +262,45 @@ static int32_t scap_write_fdlist(scap_t *handle, scap_dumper_t *d)
 	return SCAP_SUCCESS;
 }
 
+//
+// Since the process list isn't thread-safe, we at least reduce the
+// time window and write everything at once with a secondary dumper.
+// By doing so, the likelihood of having a wrong total length is lower.
+//
+scap_dumper_t *scap_write_proclist_begin(scap_t *handle)
+{
+	return scap_memstream_dump_create(handle);
+}
+int scap_write_proclist_end(scap_t *handle, scap_dumper_t *d, scap_dumper_t *d2, uint32_t totlen)
+{
+	int res = SCAP_SUCCESS;
+
+	do
+	{
+		scap_dump_flush(d2);
+
+		if(scap_write_proclist_header(handle, d, totlen) != SCAP_SUCCESS)
+		{
+			res = SCAP_FAILURE;
+			break;
+		}
+		if(scap_dump_write(d, d2->m_targetbuf, totlen) <= 0)

```

Comment:
What do you think about only passing the targetbuf itself to this function and making the `scap_dump_*` calls on d2 the responsibility of the caller?

Then the buffer to write would be a mostly generic (pointer, length) pair

Reply:
Actually I started doing it like that, but `scap_dumper_t` is opaque to `sinsp_threadinfo`, and eventually I decided to encapsulate everything and hide the details from the caller.